# Integración de Romberg

## Metodos Numéricos II

**Nombre:** Gutierrez Leon Giannine Valeria

**Fecha**: 07/10/2026

### Objetivo

Implementar el algoritmo de integracion de Romberg, correspondiente al Algoritmo 4.2 del libro de Burden como prueba del algoritmo se utilizara el ejemplo 2: $$I=\int_0^\pi \sin(x)\,dx$$

### Fundamento : Regla del trapecio y extrapolacion de Richardson

La primera aproximacion utilizada por Romberg proviene de la regla del trapecio: $$R_{1,1}=\frac{h}{2}[f(a)+f(b)]$$ donde $$h=b-a$$

Posteriormente, se van refinando los subintervalos y se obtienen nuevas aproximaciones mediante:

$$R_{i,1}=\frac{1}{2}\left[R_{i-1,1}+h\sum_{k=1}^{2^{i-2}}f\left(a+\left(k-\frac12\right)h\right)\right]$$

Una vez obtenida la nueva aproximacion trapezoidal, se realiza la extrapolacion: $$R_{i,j}=R_{i,j-1}+\frac{R_{i,j-1}-R_{i-1,j-1}}{4^{j-1}-1}$$
Despues de completar una fila, el tamaño del paso se reduce:$$h=\frac{h}{2}$$

### Algoritmo 4.2 de Burden

El algoritmo recibe como entrada los extremos $a$ y $b$ de la integral, así como un entero $n$ que indica el numero de filas que se desean calcular.

#### Paso 1
Se calcula: $$h=b-a$$ y la primera aproximacion:$$R_{1,1}=\frac{h}{2}[f(a)+f(b)]$$

#### Paso 2
Se muestra $R_{1,1}$

#### Paso 3
Para $i=2,\ldots,n$ se repiten los pasos siguientes

#### Paso 4
Se calcula una nueva aproximacion mediante la regla del trapecio refinada.

#### Paso 5
Se realizan las extrapolaciones para $j=2,\ldots,i$: $$R_{i,j}=R_{i,j-1}+\frac{R_{i,j-1}-R_{i-1,j-1}}{4^{j-1}-1}$$

#### Paso 6
Se muestran los valores de la nueva fila

#### Paso 7
Se actualiza el tamaño del paso:$$h=\frac{h}{2}$$

#### Paso 8
La nueva fila se convierte en la fila anterior para la siguiente iteracion.

#### Paso 9
El proceso termina cuando se han calculado las $n$ filas.


### Problema de prueba

Se utilizara el Ejemplo 2 de la página 160 de Burden.
Se desea calcular:
$$I=\int_0^\pi \sin(x)\,dx$$
Por lo tanto: $$a=0,\qquad b=\pi,\qquad f(x)=\sin(x)$$
El valor exacto se obtiene analiticamente: $$\int_0^\pi \sin(x)\,dx=[-\cos(x)]_0^\pi$$ $$=-\cos(\pi)-[-\cos(0)]$$ $$=1+1=2$$

Por lo tanto, el valor exacto que utilizaremos para comparar el resultado numérico es: $$\boxed{I=2}$$

In [1]:
import math #permite uzar funciones sen() y la constante pi
import numpy as np #trabaja con arreglos numéricos
import matplotlib.pyplot as plt #realiza gráficas

def f(x): #representa la función que se desea integrar
    return math.sin(x) #devuelve el valor de la función para el argumento que recibe

def romberg(f, a, b, n):
    #Paso 1: calcula h
    h=b-a
    R_anterior=[h*(f(a)+f(b))/2] # Primera aproximación trapezoidal
    tabla =[R_anterior.copy()]# Guarda las filas para mostrar la tabla
    
    #Paso 3
    for i in range(2, n + 1):   #permite repetir las operaciones para cada fila del algoritmo y genera los valores desde 2 hasta n   
        #Paso 4: calcula una nueva aproximación trapezoidal
        suma = 0.0
        for k in range(1, 2**(i-2)+1):
            x=a+(k-0.5)*h
            suma=suma+f(x) #acumula los valores de la función que aparecen en la suma del algoritmo
        R_actual = [0.5 * (R_anterior[0] + h * suma)]
        
        #Paso 5: extrapolación
        for j in range(2, i + 1):
            valor=R_actual[j-2]+(R_actual[j-2]-R_anterior[j-2])/(4**(j-1)-1)
            R_actual.append(valor)
        tabla.append(R_actual.copy()) # Guarda la fila actual
        
        #Paso 7: reduce el tamaño del paso
        h=h/2
        
        #Paso 8: actualiza la fila anterior
        R_anterior=R_actual
    
    return R_anterior, tabla
    

### Prueba del algoritmo

Según el ejemplo de Burden, podemos utilizar: $$n=6$$ porque la tabla de ej llega hasta $$R_{6,6}$$

In [6]:
a=0
b=math.pi
n=6

resultado,tabla=romberg(f, a, b, n)

print("Tabla de Romberg")

for i in range(len(tabla)):
    print(f"Fila {i+1}:", end=" ")    
    for j in range(len(tabla[i])):
        print(f"{tabla[i][j]:.8f}", end=" ")   
    print()

Tabla de Romberg
Fila 1: 0.00000000 
Fila 2: 1.57079633 2.09439510 
Fila 3: 1.89611890 2.00455975 1.99857073 
Fila 4: 1.97423160 2.00026917 1.99998313 2.00000555 
Fila 5: 1.99357034 2.00001659 1.99999975 2.00000002 1.99999999 
Fila 6: 1.99839336 2.00000103 2.00000000 2.00000000 2.00000000 2.00000000 


### Comparación con el valor exacto

In [9]:
valor_exacto=2.0
error=abs(valor_exacto-resultado[-1])

print(f"Valor exacto={valor_exacto:.12f}")
print(f"Valor aproximado={resultado[-1]:.12f}")
print(f"Error absoluto={error:.12e}")

Valor exacto=2.000000000000
Valor aproximado=2.000000000001
Error absoluto=1.321165399304e-12


Despues de aplicar el Algoritmo Integracion de Romberg al problema se obtuvieron las siguientes aproximaciones:

| Concepto | Resultado |
|:---|---:|
| Intervalo | $[0,\pi]$ |
| Funcion | $f(x)=\sin(x)$ |
| Filas | $n=6$ |
| Valor exacto | $2.000000000000$ |
| Aproximacion de Romberg | $2.000000000001$ |
| Error absoluto | $\approx 1.32\times10^{-12}$ |

##### Por lo tanto se concluye que el algoritmo implementado proporciona una aproximación muy cercana al valor exacto de la integral.
Ademas la tabla coincide con los valores mostrados en el ejemplo lo que permite comprobar el correcto funcionamiento del algoritmo